stablehlo/dialect/Base.cpp

```diff
@@ -176,5 +176,151 @@ Attribute boundsToEncoding(Attribute prototype, ArrayRef<int64_t> bounds) {
+//       Dim of lhs     Dim of rhs      Infer
+//  c0:  X              X               X
+//  c1:  X              ?               X
+//  c2:  X              ?, B(>X)        X
+//  c3:  X              ?, B(<X)        Will error out by compatible checks
+//  c4:  ?              ?               ?
+//  c5:  ?              ?, B            ?, B
+//  c6:  ?, B           ?, B            ?, B
+//  c7:  ?, B1          ?, B2           ?, min(B1, B2)
+FailureOr<std::pair<int64_t, int64_t>> inferMergedDimAndBound(
+    Optional<Location> location, int64_t dim, int64_t leftSize,
+    int64_t rightSize, int64_t leftBound, int64_t rightBound) {
+  bool isLeftStaticDim = leftSize != ShapedType::kDynamicSize;
+  bool isRightStaticDim = rightSize != ShapedType::kDynamicSize;
+  if (isLeftStaticDim && isRightStaticDim && leftSize != rightSize)
+    return emitOptionalError(location, "Mismatch dimension size ", leftSize,
+                             " and ", rightSize, " in dimension ", dim);
+  if (isLeftStaticDim || isRightStaticDim)
+    return isLeftStaticDim
+               ? std::make_pair(leftSize, ShapedType::kDynamicSize)
+               : std::make_pair(rightSize, ShapedType::kDynamicSize);
+
+  bool isLeftStaticBound = leftBound != ShapedType::kDynamicSize;
+  bool isRightStaticBound = rightBound != ShapedType::kDynamicSize;
+  if (isLeftStaticBound && isRightStaticBound)
+    return std::make_pair(ShapedType::kDynamicSize,
+                          std::min(leftBound, rightBound));
+  if (isLeftStaticBound)
+    return std::make_pair(ShapedType::kDynamicSize, leftBound);
+  if (isRightStaticBound)
+    return std::make_pair(ShapedType::kDynamicSize, rightBound);
+  return std::make_pair(ShapedType::kDynamicSize, ShapedType::kDynamicSize);
+}
+
+LogicalResult inferMostSpecificType(
+    Optional<Location> location, ValueTypeRange<ValueRange> inputTypes,
+    SmallVectorImpl<Type>& inferredReturnTypes,
+    const SmallVector<int64_t>& concatenateDims) {
+  SmallVector<RankedTensorType> rankedTypes;
+  for (auto inputType : inputTypes)
+    if (auto rankedType = inputType.dyn_cast<RankedTensorType>())
+      rankedTypes.push_back(rankedType);
+  if (rankedTypes.empty()) {
+    inferredReturnTypes.push_back(inputTypes[0]);
+    return success();
+  }
+
+  auto rank = rankedTypes[0].getRank();
+  BoundedDialectInterface* dialect = nullptr;
+  SmallVector<int64_t> inferredDimSizes(rank, ShapedType::kDynamicSize);
+  SmallVector<int64_t> inferredBounds(rank, ShapedType::kDynamicSize);
+
+  for (const auto& it : llvm::enumerate(rankedTypes)) {
+    RankedTensorType rankedType = it.value();
+    ArrayRef<int64_t> bounds;
+    if (auto boundedAttr =
+            rankedType.getEncoding().dyn_cast_or_null<BoundedAttrInterface>()) {
+      dialect = cast<BoundedDialectInterface>(&boundedAttr.getDialect());
+      bounds = boundedAttr.getBounds();
+    } else if (rankedType.getEncoding()) {
```

Comment:
nit: use a variable for encoding.

Reply:
New commit is much refined with our existing `encodingToBounds() / boundsToEncoding()` utils.